ports/stm/common-hal/sdioio/SDCard.c

```diff
@@ -0,0 +1,356 @@
+
+// XXX hard coded pin
+#define SD_DETECT_PIN GPIO_PIN_12
+#define SD_DETECT_GPIO_PORT GPIOB
+
+    /*!< Configure SD_SPI_DETECT_PIN pin: SD Card detect pin */
+    GPIO_InitStruct.Pin = SD_DETECT_PIN;
+    GPIO_InitStruct.Mode = GPIO_MODE_INPUT;
+    GPIO_InitStruct.Pull = GPIO_PULLUP;
+    HAL_GPIO_Init(SD_DETECT_GPIO_PORT, &GPIO_InitStruct);
+
+    __HAL_RCC_SDIO_CLK_ENABLE();
+
+    self->handle.Init.ClockDiv = SDIO_TRANSFER_CLK_DIV;
+    self->handle.Init.ClockEdge = SDIO_CLOCK_EDGE_RISING;
+    self->handle.Init.ClockBypass = SDIO_CLOCK_BYPASS_DISABLE;
+    self->handle.Init.ClockPowerSave = SDIO_CLOCK_POWER_SAVE_DISABLE;
+    self->handle.Init.BusWide = SDIO_BUS_WIDE_1B;
+    self->handle.Init.HardwareFlowControl = SDIO_HARDWARE_FLOW_CONTROL_DISABLE;
+    self->handle.Instance = SDIOx;
+
+    HAL_StatusTypeDef r = HAL_SD_Init(&self->handle);
+    if (r != HAL_OK) {
+        mp_raise_ValueError_varg(translate("SDIO Init Error %d"), (int)r);
+    }
+
+    HAL_SD_CardInfoTypeDef info;
+    r = HAL_SD_GetCardInfo(&self->handle, &info);
+    if (r != HAL_OK) {
+        mp_raise_ValueError_varg(translate("SDIO GetCardInfo Error %d"), (int)r);
+    }
+
+    self->num_data = 1;
+    if (num_data == 4) {
+        if ((r = HAL_SD_ConfigWideBusOperation(&self->handle, SDIO_BUS_WIDE_4B)) == HAL_SD_ERROR_NONE) {
+            DEBUG_PRINT("Switched bus to 4B mode\n");
+            self->handle.Init.BusWide = SDIO_BUS_WIDE_4B;
+            self->num_data = 4;
+        } else {
+            DEBUG_PRINT("WideBus_Enable returned %r, leaving at 1B mode\n", (int)r);
+        }
+    }
+
+    self->capacity = info.BlockNbr * (info.BlockSize / 512);
+    self->frequency = 25000000;
+
+    reserved_sdio[periph_index - 1] = true;
+
+    return;
+}
+
+uint32_t common_hal_sdioio_sdcard_get_count(sdioio_sdcard_obj_t *self) {
+    return self->capacity;
+}
+
+uint32_t common_hal_sdioio_sdcard_get_frequency(sdioio_sdcard_obj_t *self) {
+    return self->frequency; // self->frequency;
+}
+
+uint8_t common_hal_sdioio_sdcard_get_width(sdioio_sdcard_obj_t *self) {
+    return self->num_data; // self->width;
+}
+
+STATIC void check_whole_block(mp_buffer_info_t *bufinfo) {
+    if (bufinfo->len % 512) {
+        mp_raise_ValueError(translate("Buffer must be a multiple of 512 bytes"));
+    }
+}
+
+STATIC void wait_write_complete(sdioio_sdcard_obj_t *self) {
+    if (self->state_programming) {
+        HAL_SD_CardStateTypedef st = HAL_SD_CARD_PROGRAMMING;
+        // This waits up to 60s for programming to complete.  This seems like
+        // an extremely long time, but this is the timeout that micropython's
+        // implementation uses
+        for (int i=0; i < 60000 && st == HAL_SD_CARD_PROGRAMMING; i++) {
+            st = HAL_SD_GetCardState(&self->handle);
+            HAL_Delay(1);
+        };
+        self->state_programming = false;
+    }
+}
+
+STATIC void debug_print_state(sdioio_sdcard_obj_t *self, const char *what) {
```

Comment:
Again, just wondering if you want to keep this - I tend to delete most "message" debugging after a module is finished, since the usefulness of this kind of messaging rarely carries between module bugs in my experience.